### ISSUE TITLE
auto complete isn't working right 

### ISSUE THREAD

**Issue by clandestinenetworks** [eclipse/corrosion#399]
to_owned isn't coming up, don't know whats going on 
![image](https://user-images.githubusercontent.com/93053093/156328602-831c2ad0-8405-4ad0-a37a-4130a3b99f50.png)

**Comment by akurtakov** [comment]
I don't understand what the exact issue is. Would you please share a reproducer with clear explanation what is expected, link to docs if 3rd party API and etc?

**Comment by clandestinenetworks** [comment]
it obviously knows its valid so why is it not auto-completing
![Screenshot_20220302_011009](https://user-images.githubusercontent.com/93053093/156330891-6e2e9ea4-6c7e-4f9e-bd21-7637985232d8.png)
?

**Comment by akurtakov** [comment]
I get that part but in order to analyze why it happens in this case it would be easier if you share code that reproduces it so no time is lost trying to come up with example you already have.

**Comment by clandestinenetworks** [comment]
in order to analyze what happens you can simply go call 'to_owned' on a string, when you start typing to_owned if it does not appear in your autocomplete, you have successfully reproduced the issue.

**Comment by clandestinenetworks** [comment]
@akurtakov do you guys have a matrix chat or something where we could discuss a little more off topic? I'd actually be willing to dedicate as much as 40-60 hours a week of development towards getting this project working as seemlessly as what you would expect with Visual Studio or intellij. It really needs to be done. It's really going to make rust in general a more attractive option for adoption, but it's vital to me to have tools that are on par like monodevelop was until Microsoft bought Xamarin and sabotaged it because they want people to pay money for visual studio. Understandably given that I can't ever go back to C#, realistically I can't because I'm not going to pay money for tools that I've already paid thousands for over the last decade and a half. 

I need tools that work is the bottom line though, rust is a much more modern language choice and eclipse-corrosion isn't a bad start and that's what I'd like to be using. I have another option which is I could start trying to fix this nightmare: 

https://github.com/dotdevelop/dotdevelop

Since I know that I could probably be just as productive with it in most of all of my development cases; there's like a really small margin of cases in which using a language like golang or even rust specifically is vital because of its executable portability or memory management characteristics (whatever the case may be.) In a perfect world I would much prefer something like rust as a general purpose language in the sense that I have used C# as a general purpose language for years. 

That being said I have pretty high expectations for an IDE and I can possibly offer you a lot of insight as to how it can help me work faster and more efficiently.

**Comment by mickaelistria** [comment]
@clandestinenetworks Eclipse LSP4E (client support for Language Server Protocol, that rust-analyzer and Corrosion are using to communicate) does not apply any form of filtering when the Language Server returns a comptetion result that sets `isImcomplete=false` like rust-analyzer does. From the spec and the various language servers, it's not obvious what should be done in the case of isIncomplete=false. Most language server do apply filtering and sorting on their end and it works great. rust-analyzer is a bit alien here, that doesn't mean it's incorrect though (I sincerely don't know whether it is or not). I suggest that in parallel, you report the issue to rust-analyzer returning too much results for this case.
About organizational discussion that are not focused on a particular technical issue or enhancement, the best channel is to use this mailing-list: https://accounts.eclipse.org/mailing-list/corrosion-dev . This can be used to communicate the various efforts, goals and project challenges and opportunities.

**Comment by clandestinenetworks** [comment]
@mickaelistria 

Thanks, you reminded me to look at how intellij rust's completion works and it looks like their native rust analysis is proprietary and I found this:

https://github.com/intellij-rust/intellij-rust/issues/964#issuecomment-627987541

If anybody could say definitively whether LSP4E is a stretch from working correctly with rust-analyzer  it would probably be @matklad and I'll reach out to him to see if he could offer any insight in a moment.

I actually forgot that RLS was something different from rust-analyzer. Does RLS work better? From what I can gather from that post RLS doesn't offer completion for macros and rust-analyzer does. Moreover, as @matklad pointed out it for intellij that it could be potentially useful for intellij to use it's own native completion and to additionally use rust-anaylzer for macros.. so perhaps both RLS and rust-analyzer could be used for corrosion?

**Comment by mickaelistria** [comment]
> Does RLS work better?

No. RLS is kind of deprecated.

**Comment by mickaelistria** [comment]
I think LSP4E is faulty here as it simply skips filtering when isIncomplete=false and the spec doesn't make a difference about filtering for that case, so it should behave consistently independently of isIncomplete. I'm working on a fix at https://github.com/eclipse/lsp4e/pull/85

**Comment by matklad** [comment]
Specific reponses: 

* the root cause here is the LSP itself, as it doesn't provide control over who does sorting and filtering (client or server) at all. I think the way it works is that the protocol assumes that it's the client who handles relevance, but, as this doesn't work in practice, some servers put various hacks to emulate server-side relevance computations. 
* rust-analyzer currently doesn't do any server-side filtering. We should do that on day https://github.com/rust-analyzer/rust-analyzer/issues/7935.
* @mickaelistria as far as I can tell, rust-analyzer always responds with `is_incomplete: true` ([source](https://github.com/rust-analyzer/rust-analyzer/blob/a55dd29e88c6f2aa37c39b790ab7036596e74686/crates/rust-analyzer/src/handlers.rs#L815)), so I am not sure that you fix for `isIncomplete=false` would this this issue. Why do you think that "comptetion result that sets isImcomplete=false like rust-analyzer does. "? Independent of this, the fix looks good to me: as far as I understand, `isIncomplete=false` actually *requires* client-side filtering. 
* @clandestinenetworks IntilliJ-Rust analysis is not proprietary -- it's open-source, licensed under MIT, and is very friendly to contributors: https://github.com/intellij-rust/intellij-rust. 

General thoughts: 

It looks like what is happening here is that indeed the server (rust-analyzer) expects the client to do the filtering based on `filterText` and that the client doesn't do that. Haven't looked in details though.

**Comment by mickaelistria** [comment]
@matklad thanks, particularly for the link to https://github.com/rust-analyzer/rust-analyzer/issues/7935 . I'm glad to see that corrosion and rust-analyzer share the same feelings regarding this issue.
My comment earlier was wrong, I was talking about the case of `isIncomplete=true`, that makes LSP4E skip filtering. https://github.com/eclipse/lsp4e/pull/85 is supposed to fix that and make LSP4E behave closer to other clients here.

**Comment by clandestinenetworks** [comment]
@matklad thanks for the clarification but I didn't mean proprietary in the licensing sense of the word I meant that specifically the completion engine in intellij rust is a one-off solution that nobody else uses because it was written to adapt the libraries that intellij uses for all of its products. Also can you confirm whether or not Intellij Rust CE debugging is possible? The debug button is just grayed out on mine and I can't find any documentation that explains how to enable it.

I found a product matrix somewhere that was hidden away that seemed to indicate that debugging was not available in CE and that you would have to purchase Idea Ultimate in order to debug. I'd really like to believe this is not true but I have not found any documentation or evidence that would suggest otherwise =)

**Comment by matklad** [comment]
@clandestinenetworks a better place to ask general Rust dev-tooling questions would be the IDE topic on the Rust forum: https://users.rust-lang.org/c/ide/14. Regarding debugging, yes, at the moment for Rust that is only available in paid products like CLion or IntelliJ Ultimate.

**Comment by clandestinenetworks** [comment]
@matklad of course I could, if I wanted a direct answer but the fact that I can't simply find the answer for myself about something that realistically anybody would expect from an integrated development environment is pretty telling. It's actually infuriating and it's more or less the reason why I've arrived at corrosion; I need something that does what it's supposed to do and isn't a trick to try to rope me into spending more money on intellij products. I think people who can't really afford intellij deserve a more comprehensive means than nano editor and VS.code as well. Frankly it's very frustrating to have an intention to build something that interfaces with the real world but can't because you have to listen to people condescend to you about whether nano editor is better than emacs and what "more than you need" means. 

Not to mention that my opinion on languages such as rust and C# which are syntactically complex enough that it is counter-productive to NOT have reliable intellisense, unless you fancy yourself particularly good at buckshot development and 100+ rust API documentation browser tabs for things that you might have gleaned from good intellisense.

**Comment by mickaelistria** [comment]
I'm removing all the messages that discuss IJ or other products strategies, as they are not really helpful for resolution of this issue, nor for Corrosion development in general.

**Comment by mickaelistria** [comment]
@clandestinenetworks Try installing latest LSP4E build from http://download.eclipse.org/lsp4e/snapshots/ which should include a fix for this issue.

**Comment by clandestinenetworks** [comment]
@mickaelistria looks good and thank you. Also I spent this morning figuring out how to build corrosion on my dev box. I just installed the plugin directly as you suggested but just so that I understand is LSP4E included in the build for corrosion? 
I have a couple of other points I'd like to bring up here about completion and I'll try to show you what I mean. 

- I can see that to_owned is definitely showing up a bit better, however in the case that I've typed `to_ow` we can pretty much rule out that I would think that `to_ascii_lowercase` and `to_lowercase` wouldn't be in the list at all but at the very least they would not be the first items in the list:

![Screenshot_20220302_142958](https://user-images.githubusercontent.com/93053093/156461169-b2758f45-f644-4874-b49f-182e4a6d9c96.png)

- ordinarily when I try to complete a struct I expect to see members of the struct being first in the list however currently they are just sorted alphabetically amongst the rest of the list items: 
![Screenshot_20220302_144003](https://user-images.githubusercontent.com/93053093/156462345-9fe1d7c8-0812-4288-bb28-677fcda35b49.png)

You'll have to bare with me on this because this example is taken from a C# IDE but: 

![Screenshot_20220302_144755](https://user-images.githubusercontent.com/93053093/156462838-73c31ccf-46eb-4f56-b7ad-790ae9d11b48.png)
 
all of the standard members of IDictionary are shown first, I believe it does have an order of precedence in which members of a class are shown first (before interface members.) 

And then as you scroll down the list to the bottom you will see all of the short options (for expanding it to a for loop and so forth) are all the way at the bottom: 

![Screenshot_20220302_145010](https://user-images.githubusercontent.com/93053093/156463130-6453e480-44bd-44f1-a2b1-931e2a9cb502.png)

- With Ok/Err, I actually had to press control space here to get this to pop, but that's unsurprising since there probably isn't anything for it to suggest currently however, based on the signature of the function to which Err should apply; can it not

**Comment by clandestinenetworks** [comment]
Another thing I've been noticing that is screwing me up is when you do new like from this list:
![Screenshot_20220302_152124](https://user-images.githubusercontent.com/93053093/156466361-b2283d2f-13e6-431a-a7ae-de04b74fef64.png)

it throws in this arbitrary variable 'x' which isn't even defined anywhere in my scope:
![Screenshot_20220302_152230](https://user-images.githubusercontent.com/93053093/156466485-8f0d6d4d-a647-4716-9107-32afbd26e59f.png)

Here's what I would do, rather than inserting variables, get the signature of the fn new and show a hint: 

![Screenshot_20220302_152425](https://user-images.githubusercontent.com/93053093/156466787-ce03ae9d-fb10-4d7e-a006-aa16dff4b7a9.png)

that's really useful to me, it's saved my adhd mind a ton of time. 

Another one that goes hand in hand with that is it would be nice to have a shortcut like `///` which when typed automatically inserts a skeleton docblock. For C# the autogenerated docstring you get might look something like:

![Screenshot_20220302_152932](https://user-images.githubusercontent.com/93053093/156467156-735b2f00-6a84-4fa1-bea9-621bc4efcea7.png)

and when you enter a description for the params, that description will actually show up in the function variable hint when attempting to make a function call, for example if you have 

![Screenshot_20220302_153148](https://user-images.githubusercontent.com/93053093/156467393-df469293-011f-42e0-859f-6424685d29ad.png)

then: 
![Screenshot_20220302_153245](https://user-images.githubusercontent.com/93053093/156467460-14b842e8-e30a-4e1e-88f3-5d408ca56ab5.png)

**Comment by clandestinenetworks** [comment]
@mickaelistria sorry if that was too much but I really want to try to figure out what your guys plan for improvements are and whether or not you have any interest. Given the complexity that I know is involved in building expression evaluators and code generators I'd probably also like to run this by @matklad and see what he thinks about it since rust-analyzer seems to be the thing to do but I have to wonder how he feels about it given that 2 years ago he didn't seem entirely confident that rust-analyzer would amount to anything more than something experimental. And that's totally fair, I just want to make sure. 


I'm 99% sure myself that if this isn't in anybody's interest that I (really don't want to) but will fork whatever I need to in order to accomplish what I would like to see done. I think there really needs to be a long term, reliable IDE for rust that I can always use productively and that I can teach others how to use productively that doesn't get completely derailed by having to search out and install a dozen different packages, just a sane set of tools in one single distribution that (within reason) are always the same that work the way that I expect them to (best effort.) 

VS.code has proven to be the one editor for everything for which nothing works consistently and the things that people say works don't. I don't want to mess with a market place or an app store. I want to install rust-ide and that's it. I'm a really big fan that you added the ability to install rust / rust up and to install rust-analyzer from within the ide itself; that's good because I don't want to install those things or remember how to. In fact I would much prefer if it just automatically did it, and left the option open to change to using some other toolchain somewhere else in the userland manually if the case needs to be. 

If you're not interested, can you at least understand? Is that not inherently what integrated development environment means? I don't think too many people do and I should know because I've used everything from Turbo C, Metrowerks Codewarrior, Visual Studio.. uh I can't even imagine how you would emulate the kind of modality that you need to render the kind of intellisense I've shown you so far.. 

I do have trouble recalling key bindings when I'm actually writing code so it's nice to have a visual representation of a start / stop / step debugging / set breakpoints too that I still am not seeing in spacemacs. More importantly though spacemacs isn't something that I can teach people how to use or that I can reliably use collaboratively.

**Comment by clandestinenetworks** [comment]
I can definitely appreciate but I can't take it seriously.. for a number of reasons. I wish people would stop telling me that they use neovim and that it's "good enough" because I don't care 
![Screenshot_20220302_160316](https://user-images.githubusercontent.com/93053093/156470456-05310c33-6f3a-4b64-9968-f65716f08aa5.png)

**Comment by mickaelistria** [comment]
@clandestinenetworks What gets returned and inserted by the completion is part of rust-analyzer and issues should be reported there. Corrosion just uses rust-analyzer and doesn't own nor maintain the content of completion. The only thing that are on the client side is mostly filtering and sorting, that can still be improved; and the "snippets" which would actually also probably more useful on server-side (I'm about to open a PR to at least reduce their priority). For those issues on client-side, please open separate dedicated issues and we'll investigate.
But really, it looks like most of the issues you face would require development and contribution to rust-analyzer, which should be a more comfortable project for you to work with if you're familiar with Rust, and also a more "profitable" project as improvements in rust-analyzer would cascade to all IDEs/Editors at once.
